core/src/test/java/com/datastax/oss/driver/When.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver;
+
+import com.datastax.oss.driver.shaded.guava.common.collect.Collections2;
+import com.datastax.oss.driver.shaded.guava.common.collect.ImmutableList;
+import com.datastax.oss.driver.shaded.guava.common.collect.ImmutableMap;
+import java.util.Iterator;
+import java.util.List;
+
+public class When<T> implements Iterable<Iterable<Object>> {
```

Comment:
Where it is plugin into logic? I don't see any usage of this new class besides the test

Reply:
Yeah, it's only a testing thing.  I added it to make the data provider definition more explicitly declarative.

Reply:
To be honest, I'm not a fan. To me, it doesn't make the provider declarations any easier to read.

If the goal is to document what each field in the provided data is (and I agree it's a weakness in the array syntax), then I find that a simple comment at the beginning does a pretty decent job. See example [here](https://github.com/datastax/java-driver/blob/4.x/query-builder/src/test/java/com/datastax/oss/driver/api/querybuilder/BuildableQueryTest.java#L42).

Reply:
I personally don't find that example nearly as declarative (or as readable) as what's in this PR, but at the same time I don't want to introduce an impl that's only going to be used in some subset of unit tests.  I'll see if I can re-work my test into a middle ground between those ideas (and remove the "when" impl) but it does seem to me like there's _considerable_ room to improve how we define these providers in the future.